exp/query/cache.go

```diff
@@ -0,0 +1,163 @@
+)
+
+const (
+	evictFraction = 0.1 // When evicting, remove 10% of max disk capacity.
+	CachePrefix   = "weaviate-cache"
+)
+
+// DiskCache is LRU disk based cache for tenants on-disk data.
+type DiskCache struct {
+	// key: <collection>/<tenant>
+	tenants   map[string]*list.Element
+	evictList *list.List
+	mu        sync.Mutex // TODO(kavi): Make it more fine-grained than global mutex
+
+	// more than this will start evicting the tenants
+	// maxCap in bytes.
+	maxCap int64
+
+	basePath string
+}
+
+func NewDiskCache(basePath string, maxCap int64) *DiskCache {
+	return &DiskCache{
+		tenants:   make(map[string]*list.Element),
+		evictList: list.New(),
+		basePath:  path.Join(basePath, CachePrefix),
+		maxCap:    maxCap,
+	}
+}
+
+// Assumes id all small-case
+func (d *DiskCache) Tenant(collection, tenantID string) (*TenantCache, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if c, exists := d.tenants[d.TenantKey(collection, tenantID)]; exists {
+		d.evictList.MoveToFront(c)
+		c.Value.(*TenantCache).LastAccessed = time.Now()
+		return c.Value.(*TenantCache), nil
+	}
+	return nil, ErrTenantNotFound
+}
+
+// AddTenant is called after having the files in right directory
+// AddTenant checks if file is present on that directory
+// deterministically generated from `collection` & `tenant` with basePath
+func (d *DiskCache) AddTenant(collection, tenantID string, version uint64) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	tc := &TenantCache{
+		TenantID:     tenantID,
+		Collection:   collection,
+		Version:      version,
+		LastAccessed: time.Now(),
+		basePath:     d.basePath,
+	}
+
+	_, err := os.Stat(tc.AbsolutePath())
+	if err != nil && errors.Is(err, os.ErrNotExist) {
+		return errors.Join(ErrTenantDirectoryFound, err)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to get local path of the tenant: %w", err)
+	}
+
+	c := d.evictList.PushFront(tc)
+	d.tenants[d.TenantKey(tc.Collection, tc.TenantID)] = c
+
+	// evit if the size is getting filled up
+	// TODO(kavi): Worth doing it on different go routine? But will get complex with correctness
+	if d.usage() > d.maxCap {
+		if err := d.evict(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BasePath returns base directory of cache data.
+// Useful for client to know where to copy the files from upstream when using
+// the cache.
+func (d *DiskCache) BasePath() string {
+	return d.basePath
+}
+
+// return total current usage of disk cache in bytes
+func (d *DiskCache) usage() int64 {
+	var sizeBytes int64
+
+	filepath.Walk(d.basePath, func(path string, info os.FileInfo, err error) error {
+		if err == nil && !info.IsDir() {
+			sizeBytes += info.Size()
+		}
+		return nil
+	})
+	return sizeBytes
+}
+
+func (d *DiskCache) evict() error {
+	evictSize := int64(float64(d.maxCap) * evictFraction)
+
+	for d.usage() > (d.maxCap - evictSize) {
+		e := d.evictList.Back()
+		if e == nil {
+			// NOTE: Shouldn't happen.
```

Comment:
NIT: Maybe we should log a big error that this shouldn't happen but it is happening ? What could ops do if such a warning appears ?

Reply:
good thinking 👍 

I would rather panic here. This means basically our querier cannot store any copy of tenant data, which intern means it cannot serve any query. 

Once we take it stage setup, we can learn from it and see if we can do any workaround. For now I want it to fail immediately to stop making it worse.

Reply:
Done